File: Baekjoon/4195친구네트워크.cpp
#include<iostream>
#include<map>
#include<string>
using namespace std;

int parent[200002];
int set[200002];

int getParent(int n) {
	if (parent[n] == n) {
		return n;
	}
	return parent[n] = getParent(parent[n]);
}

int unionParent(int a, int b) {
	int parentA = getParent(a);
	int parentB = getParent(b);
	if (parentA < parentB) {
		parent[parentB] = parentA;
		return set[parentA] += set[parentB];
	}
	else if(parentA > parentB) {
		parent[parentA] = parentB;
		return set[parentB] += set[parentA];
	}
	return set[parentA];
}

int main() {
	ios_base::sync_with_stdio(false), cin.tie(NULL), cout.tie(NULL);
	int testcase;
	cin >> testcase;
	while (testcase--) {
		int n, num= 0;
		map<string, int> people;
		string people1, people2;
		cin >> n;
		for (int i = 1; i <= 2 * n; i++) {
			parent[i] = i;
			set[i] = 1;
		}
		while (n--) {
			cin >> people1 >> people2;
			if (people.find(people1) == people.end()) {
				people[people1] = ++num;
			}
			if (people.count(people2) == 0) {
				people[people2] = ++num;
			}
			cout << unionParent(people[people1], people[people2]) << '\n';
		}
		people.clear();
	}
	return 0;
}

File: Baekjoon/4195친구네트워크.py
from collections import defaultdict

def getParent(a):
    if not parent[a]:
        parent[a] = a
        return a
    if a == parent[a]:
        return a
    else:
        parent[a] = getParent(parent[a])
        return parent[a]

def union(a, b):
    a = getParent(a)
    b = getParent(b)
    if a == b:
        return count[a]
    parent[b] = a
    count[a] += count[b]
    return count[a]

t = int(input())

while t:
    t -= 1
    n = int(input())
    parent = defaultdict(bool)
    count = defaultdict(lambda: 1, {})
    for _ in range(n):
        a, b = input().split()
        print(union(a, b))
